File: python3/l0020_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        open_close_map = {'(': ')', '[': ']', '{': '}'}
        stack = []
        for c in s:
            if c in '([{':
                stack.append(c)
            else:
                if not stack or open_close_map[stack.pop()] != c:
                    return False
        return len(stack) == 0

File: java/src/main/java/leetcode/L0020ValidParentheses.java
package leetcode;

import java.util.*;

class L0020ValidParentheses {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            Character c = s.charAt(i);
            if (c == '[') {
                stack.push(']');
            } else if (c == '(') {
                stack.push(')');
            } else if (c == '{') {
                stack.push('}');
            } else {
                if (stack.empty()) {
                    return false;
                }
                if (c != stack.pop()) {
                    return false;
                }
            }
        }
        return stack.empty();
    }
}